src/nickelclock.cc

```diff
@@ -242,27 +214,96 @@ void NC::setFooterStylesheet(ReadingFooter *rf)
     auto l = rf->layout();
     if (origFooterMargin < 0)
         origFooterMargin = l->contentsMargins().left();
-    int newMargin = settings.hMargin();
+    int newMargin = settings.margin();
     if (newMargin < 0)
         newMargin = origFooterMargin / 10;
     QString s = QStringLiteral("qproperty-footerMargin: %1;").arg(newMargin);
     QString ss = origFooterStylesheet;
     rf->setStyleSheet(ss.replace(footerMarginRe, s));
 }
 
+TimeLabel* NC::createTimeLabel()
+{
+    TimeLabel *tl = (TimeLabel*) ::operator new (128); // Actual size 88 bytes
+    TimeLabel__TimeLabel(tl, nullptr);
+    tl->setObjectName(nc_widget_name);
+    auto hAlign = settings.clockPosition() == Left ? Qt::AlignLeft : Qt::AlignRight;
+    tl->setAlignment(hAlign | Qt::AlignVCenter);
+    tl->setStyleSheet(ncLabelStylesheet());
+    return tl;
+}
+
+QWidget* NC::createBatteryWidget()
+{
+    BatteryType type = settings.batteryType();
+    QWidget *battery = new QWidget();
+    QHBoxLayout *l = new QHBoxLayout();
+    NCBatteryLabel *level = nullptr;
+    N3BatteryStatusLabel *icon = nullptr;
+
+    if (type == Level || type == Both) {
+        int initLevel = getBatteryLevel();
+        level = new NCBatteryLabel(initLevel, settings.batteryLabel());
+        level->setStyleSheet(ncLabelStylesheet());
+        l->addWidget(level, 0, Qt::AlignVCenter);
+    }
+
+    if (type == Icon || type == Both) {
+        icon = (N3BatteryStatusLabel*) ::operator new (256); // Actual size 208 bytes
+        N3BatteryStatusLabel__N3BatteryStatusLabel(icon, nullptr);
+        l->addWidget(icon, 0, Qt::AlignVCenter);
+    }
+
+    l->setContentsMargins(0, 0, 0, 0);
+    battery->setLayout(l);
+    battery->setStyleSheet("padding: 0px; margin: 0px; background-color: transparent;");
+    battery->show();
+    return battery;
+}
+
+// Trying to get the battery level out of Nickel seems to be more trouble
+// than it's worth, therefore get it via sysfs
+int NC::getBatteryLevel()
+{
+    int battery = 100;
+    QFile bcFile;
+    bcFile.setFileName(QFile::exists(nc_sysfs_gen_bat_cap) 
+                        ? nc_sysfs_gen_bat_cap
+                        : nc_sysfs_mc13892_bat_cap);
+    if (bcFile.open(QIODevice::ReadOnly | QIODevice::Text)) {
+        bool ok = false;
+        battery = bcFile.readAll().trimmed().toInt(&ok);
+        if (!ok) {
+            nh_log("failed to get battery level");
+            battery = 100;
+        }
+    }
+    return battery;
+}
```

Comment:
@NiLuJe do you see any issues here?

Reply:
Not entirely familiar with `QFile`, so I'm not sure if that's necessary here, but I don't see it getting `close`d?

Reply:
(Possibly cache the `exists` test result on startup, as it's never going to change at runtime ;)).

Reply:
I thought I read in the documentation that QFile closes on destruction, but I just reread the documentation, and it doesn't say anything about it so I probably do need to close it. Good catch thank you.

Reply:
Oh, hang on, yes it does:

> [virtual]QFile::~QFile()
> Destroys the file object, closing it if necessary.